File: src/main/org/nlogo/api/ValueConstraint.java
package org.nlogo.api;

/**
 * Interface for objects which provide constraints for values.
 */
public interface ValueConstraint {

  /**
   * Throws a Violation condition if the input is not acceptable.
   */
  void assertConstraint(Object val) throws Violation, LogoException;

  /**
   * Returns the constrained value, which can differ from the input.
   * Throws a Violation condition if the input is not coercable.
   */
  Object coerceValue(Object val) throws Violation, LogoException;

  /**
   * Returns the default value for this constraint
   */
  Object defaultValue();

  //void updateConstraint() throws CompilerException, LogoException ;

  class Violation extends AgentException {
    public Violation(String message) {
      super(message);
    }
  }
}

File: src/main/api/ValueConstraint.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.api

object ValueConstraint {
  class Violation(message: String) extends AgentException(message)
}

/** Interface for objects which provide constraints for values */
trait ValueConstraint {

  /** Throws a Violation condition if the input is not acceptable. */
  @throws(classOf[ValueConstraint.Violation])
  def assertConstraint(value: AnyRef)

  /** Returns the constrained value, which can differ from the input.
    * Throws a Violation condition if the input is not coercable. */
  @throws(classOf[ValueConstraint.Violation])
  def coerceValue(value: AnyRef): AnyRef

  /** Returns the default value for this constraint. */
  def defaultValue: AnyRef
}
